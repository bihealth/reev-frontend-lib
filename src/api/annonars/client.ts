import { chunks } from '@reactgular/chunks'

import type { LinearStrucvar, Seqvar } from '../../lib/genomicVars'
import { urlConfig } from '../../lib/urlConfig'
import { ClinvarPerGeneRecord } from '../../pbs/annonars/clinvar/per_gene'
import { Record as GeneInfoRecord } from '../../pbs/annonars/genes/base'
import {
  ClinvarSvQueryResponse,
  GeneInfoResult,
  GeneSearchResponse,
  SeqvarInfoResponse
} from './types'

/**
 * Client for accessing annonars REST API.
 */
export class AnnonarsClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws Error if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlAnnonars !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlAnnonars
    } else {
      throw new Error('Configuration error: API base URL not configured')
    }
  }

  /**
   * Fetch gene information via annonars REST API.
   *
   * @param hgncId HGNC ID, e.g., `"HGNC:26467"`.
   */
  async fetchGeneInfo(hgncId: string): Promise<GeneInfoResult> {
    const response = await fetch(`${this.apiBaseUrl}/genes/info?hgnc_id=${hgncId}`, {
      method: 'GET'
    })
    return await response.json()
  }

  /**
   * Fetch variant information via annonars and mehari REST APIs.
   *
   * @param seqvar The variant to retrieve the information for.
   */
  async fetchVariantInfo(seqvar: Seqvar): Promise<SeqvarInfoResponse> {
    const { genomeBuild, chrom, pos, del, ins } = seqvar
    let chromosome = chrom.replace('chr', '')
    if (genomeBuild !== 'grch37') {
      chromosome = `chr${chrom}`
    }

    const url =
      `${this.apiBaseUrl}/annos/variant?genome_release=${genomeBuild}&` +
      `chromosome=${chromosome}&pos=${pos}&reference=${del}&` +
      `alternative=${ins}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch variant info: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return SeqvarInfoResponse.fromJson(responseJson)
  }

  /**
   * Fetch per-gene ClinVar information via annonars REST API.
   *
   * @param hgncId
   * @returns
   */
  async fetchGeneClinvarInfo(hgncId: string): Promise<ClinvarPerGeneRecord> {
    const response = await fetch(`${this.apiBaseUrl}/genes/clinvar?hgnc_id=${hgncId}`, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch gene clinvar info: ${response.statusText}`)
    }
    return await response.json()
  }

  /**
   * Search for genes, e.g., by symbol, via annonars REST API.
   *
   * @param query Query string to search for.
   * @returns Promise with gene search response.
   * @throws Error if the request fails.
   */
  async fetchGenes(query: string): Promise<GeneSearchResponse> {
    const response = await fetch(`${this.apiBaseUrl}/genes/search?q=${query}`, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch genes: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return GeneSearchResponse.fromJson(responseJson)
  }

  /**
   * Fetch gene informations via annonars REST API.
   *
   * @param hgncIds Array of HGNC IDs to use, e.g., `["HGNC:26467"]`.
   * @param chunkSize How many IDs to send in one request.
   * @returns Promise with an array of gene information objects.
   */
  async fetchGeneInfos(hgncIds: string[], chunkSize?: number): Promise<GeneInfoRecord[]> {
    const hgncIdChunks = chunks(hgncIds, chunkSize ?? 10)

    const promises = hgncIdChunks.map(async (chunk) => {
      const url = `${this.apiBaseUrl}/genes/info?hgnc_id=${chunk.join(',')}`

      const response = await fetch(url, {
        method: 'GET'
      })
      if (!response.ok) {
        throw new Error(`failed to fetch gene infos: ${response.statusText}`)
      }
      return response
    })

    const responses = await Promise.all(promises)
    const resultJsons = await Promise.all(responses.map((response) => response.json()))

    const result: GeneInfoRecord[] = []
    resultJsons.forEach((chunk: any) => {
      for (const value of Object.values(chunk.genes)) {
        // @ts-ignore
        result.push(GeneInfoRecord.fromJson(value as JsonValue))
      }
    })
    return result
  }

  /**
   * Fetch overlapping ClinVar strucvars via annonars REST API.
   */
  async fetchClinvarStrucvars(
    strucvar: LinearStrucvar,
    pageSize: number = 1000,
    minOverlap: number = 0.1
  ): Promise<ClinvarSvQueryResponse> {
    const { genomeBuild, chrom, start, stop } = strucvar
    const url =
      `${this.apiBaseUrl}/clinvar-sv/query?genomeRelease=${genomeBuild}&` +
      `chromosome=${chrom}&start=${start}&stop=${stop}&pageSize=${pageSize}&` +
      `minOverlap=${minOverlap}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch clinvar strucvars: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return ClinvarSvQueryResponse.fromJson(responseJson)
  }
}
