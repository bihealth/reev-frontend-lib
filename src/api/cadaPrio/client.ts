import { urlConfig } from '../../lib/urlConfig'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { Response } from './types'

/**
 * Client for the CADA-Prio API.
 */
export class CadaPrioClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws ConfigError if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlCadaPrio !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlCadaPrio
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Predict similarity of genes with list of HPO terms.
   *
   * @param hpoTerms HPO term IDs (e.g. `HP:0000001`)
   * @param geneSymbols Gene symbols (e.g. `BRCA1`)
   * @returns Promise with response of impact prediction.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async predictGeneImpact(hpoTerms: string[], geneSymbols?: string[]): Promise<Response> {
    const geneSuffix = geneSymbols ? `&gene_symbols=${geneSymbols.join(',')}` : ''
    const url = `${this.apiBaseUrl}/api/v1/predict?hpo_terms=${hpoTerms.join(',')}${geneSuffix}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(`failed to fetch gene impact prediction: ${response.statusText}`)
    }
    try {
      const responseJson = await response.json()
      return Response.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse gene impact prediction response: ${e}`)
    }
  }
}
