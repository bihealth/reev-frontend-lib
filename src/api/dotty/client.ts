import { urlConfig } from '../../lib/urlConfig'
import { SpdiResult, TranscriptResult } from './types'

export class DottyClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws Error if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlDotty !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlDotty
    } else {
      throw new Error('Configuration error: API base URL not configured')
    }
  }

  /**
   * Convert the given query to SPDI notation.
   *
   * @param q Query to conert.
   * @param assembly Assembly to use.
   * @returns Promise with the response form dotty or null if the request failed.
   */
  async toSpdi(q: string, assembly: 'GRCh37' | 'GRCh38' = 'GRCh38'): Promise<SpdiResult | null> {
    const escapedQ = encodeURIComponent(q)
    const url = `${this.apiBaseUrl}/api/v1/to-spdi?q=${escapedQ}&assembly=${assembly}`
    const response = await fetch(url, {
      method: 'GET'
    })
    if (response.status == 200) {
      const responseJson = await response.json()
      return SpdiResult.fromJson(responseJson)
    } else {
      return null
    }
  }

  /**
   * Fetch transcripts for the given HGNC ID.
   *
   * @param hgncId HGNC ID to fetch transcripts for.
   * @param assembly Assembly to use.
   * @returns Promise with the response form dotty or null if the request failed.
   */
  async fetchTranscripts(
    hgncId: string,
    assembly: 'GRCh37' | 'GRCh38' = 'GRCh38'
  ): Promise<TranscriptResult | null> {
    const url = `${this.apiBaseUrl}/api/v1/find-transcripts?hgnc_id=${hgncId}&assembly=${assembly}`
    const response = await fetch(url, {
      method: 'GET'
    })
    if (response.status == 200) {
      const responseJson = await response.json()
      return TranscriptResult.fromJson(responseJson)
    } else {
      return null
    }
  }
}
