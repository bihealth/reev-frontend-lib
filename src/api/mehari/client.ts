import type { Seqvar, Strucvar } from '../../lib/genomicVars'
import { urlConfig } from '../../lib/urlConfig'
import { GeneTranscriptsResponse } from '../../pbs/mehari/server'
import { GenomeBuild } from '../../pbs/mehari/txs'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { SeqvarResult, StrucvarResult } from './types'

/**
 * Client for Mehari API requests.
 */
export class MehariClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws ConfigError if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlMehari !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlMehari
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Retrieve consequences for sequence variants.
   *
   * @param seqvar Sequence variant to retrieve consequences for.
   * @param hgncId HGNC ID of gene to restrict results to.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveSeqvarsCsq(seqvar: Seqvar, hgncId?: string): Promise<SeqvarResult> {
    const { genomeBuild, chrom, pos, del, ins } = seqvar
    const hgncSuffix = hgncId ? `&hgnc_id=${hgncId}` : ''
    const url =
      `${this.apiBaseUrl}/api/v1/seqvars/csq?genome_release=${genomeBuild}&` +
      `chromosome=${chrom}&position=${pos}&reference=${del}&` +
      `alternative=${ins}${hgncSuffix}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(
        `Failed to fetch sequence variant consequences: ${response.statusText}`
      )
    }
    try {
      const responseJson = await response.json()
      return SeqvarResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(
        `Failed to parse sequence variant consequences response: ${e}`
      )
    }
  }

  /**
   * Retrieve consequences for structural variants.
   *
   * @param strucvar Structural variant to retrieve consequences for.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveStrucvarsCsq(strucvar: Strucvar): Promise<StrucvarResult> {
    let url: string
    const { svType, genomeBuild, chrom, start } = strucvar
    if (svType === 'BND' || svType === 'INS') {
      url =
        `${this.apiBaseUrl}/api/v1/strucvars/csq?genome_release=${genomeBuild}&` +
        `chromosome=${chrom}&start=${start}&stop=${start + 1}&sv_type=${svType}`
    } else {
      const { stop } = strucvar
      url =
        `${this.apiBaseUrl}/api/v1/strucvars/csq?genome_release=${genomeBuild}&` +
        `chromosome=${chrom}&start=${start}&stop=${stop}&sv_type=${svType}`
    }

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(
        `Failed to fetch structural variant consequences: ${response.statusText}`
      )
    }
    try {
      const responseJson = await response.json()
      return StrucvarResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(
        `Failed to parse structural variant consequences response: ${e}`
      )
    }
  }

  /**
   * Retrieve transcripts given a HGNC ID.
   *
   * @param hgncId HGNC ID of gene to retrieve transcripts for.
   * @param genomeBuild Genome build to restrict results to.
   * @param pageSize Number of results to return per page.
   * @param nextPageToken Token to retrieve the next page of results.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveGeneTranscripts(
    hgncId: string,
    genomeBuild: GenomeBuild,
    pageSize: number = 1000,
    nextPageToken?: string
  ): Promise<GeneTranscriptsResponse> {
    const urlGenomeBuild = genomeBuild === GenomeBuild.GENOME_BUILD_GRCH37 ? 'grch37' : 'grch38'
    const url =
      `${this.apiBaseUrl}/api/v1/genes/transcripts?hgnc_id=${hgncId}&` +
      `genome_build=${urlGenomeBuild}&page_size=${pageSize}` +
      (nextPageToken ? `&next_page_token=${nextPageToken}` : '')

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(`Failed to fetch transcripts: ${response.statusText}`)
    }
    try {
      const responseJson = await response.json()
      return GeneTranscriptsResponse.fromJson(responseJson, { ignoreUnknownFields: true })
    } catch (e) {
      throw new InvalidResponseContent(`Failed to parse transcript response: ${e}`)
    }
  }
}
