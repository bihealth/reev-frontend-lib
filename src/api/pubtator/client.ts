import { urlConfig } from '../../lib/urlConfig'
import type { SearchResult } from './types'

/**
 * Client for PubTator V3 API queries.
 */
export class PubtatorClient {
  /** API base URL to use in this client instance. */
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlPubtator !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlPubtator
    } else {
      throw new Error('Configuration error: API base URL not configured')
    }
  }

  /**
   * Perform search for the given HGNC symbol.
   *
   * @param hgncSymbol HGNC symbol to search for.
   * @returns Promise for the search results.
   * @throws Error if the search fails.
   */
  async performSearch(hgncSymbol: string): Promise<{ [key: string]: SearchResult }> {
    const url = `${this.apiBaseUrl}/search/?text=@GENE_${hgncSymbol}`
    const searchRes = await fetch(url, {
      method: 'GET'
    })
    if (!searchRes.ok) {
      throw new Error(`Error running PubTator 3 search: ${searchRes.statusText}`)
    }
    const searchData = await searchRes.json()

    // Then, extract PMID list and retrieve biocjson for the PMIDs
    const pmids: string[] = searchData!.results!.map((doc: any) => doc.pmid)
    const exportRes = await fetch(
      `${this.apiBaseUrl}/publications/export/biocjson` + `?pmids=${pmids.join(',')}`
    )
    if (!exportRes.ok) {
      throw new Error(`Error running PubTator 3 export: ${exportRes.statusText}`)
    }
    const exportDataText = await exportRes.text()
    const exportDataLines = exportDataText.split(/\n/)

    // Zip search results and exports into searchResults
    const searchResults: { [key: string]: SearchResult } = {}
    for (const searchDataRecord of searchData.results) {
      searchResults[searchDataRecord.pmid] = {
        searchResult: searchDataRecord,
        abstract: undefined
      }
    }
    for (const exportDataLine of exportDataLines) {
      if (exportDataLine) {
        const exportDataRecord = JSON.parse(exportDataLine)
        searchResults[exportDataRecord.pmid].abstract = exportDataRecord
      }
    }

    return searchResults
  }
}
