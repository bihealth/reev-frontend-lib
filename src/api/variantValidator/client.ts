import type { Seqvar } from '../../lib/genomicVars'
import { urlConfig } from '../../lib/urlConfig'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { Response as VariantValidatorResponse } from './types'

/**
 * Client for the VariantValidator API.
 */
export class VariantValidatorClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws ConfigError if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlVariantValidator !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlVariantValidator
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Call variant validator API via proxy.
   *
   * @param seqvar The `Seqvar` object to be validated.
   * @returns The response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async fetchVvResults(seqvar: Seqvar): Promise<VariantValidatorResponse> {
    const { genomeBuild, chrom, pos, del, ins } = seqvar
    const release = genomeBuild === 'grch37' ? 'hg19' : 'hg38'
    const url =
      `${this.apiBaseUrl}/${release}/` +
      `${chrom}-${pos}-${del}-${ins}/mane/?content-type=application%2Fjson`

    const response = await fetch(url, { method: 'GET' })
    if (!response.ok) {
      throw new StatusCodeNotOk(`Failed to fetch ACMG rating for ${seqvar.userRepr}`)
    }
    try {
      const responseJson = await response.json()
      return VariantValidatorResponse.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`Invalid response content from VariantValidator: ${e}`)
    }
  }
}
