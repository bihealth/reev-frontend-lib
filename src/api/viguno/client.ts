import { urlConfig } from '../../lib/urlConfig'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { HpoGenesResult, HpoOmimsResult, HpoTermResult } from './types'

export class VigunoClient {
  private apiBaseUrl: string

  /**
   * Construct Viguno client.
   *
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws Error if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlViguno !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlViguno
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Resolve OMIM term by its ID.
   *
   * @param id OMIM ID to resolve.
   * @returns Promise with the response from the server.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async resolveOmimTermById(id: string): Promise<HpoOmimsResult> {
    const url = `${this.apiBaseUrl}/api/v1/hpo/omims?omim_id=${id}`
    const response = await fetch(url, {
      method: 'GET'
    })

    if (!response.ok) {
      const errorBody = await response.json()
      throw new StatusCodeNotOk(errorBody.msg || response.statusText)
    }

    try {
      const responseJson = await response.json()
      return HpoOmimsResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse OMIM response: ${e}`)
    }
  }

  /**
   * Query OMIM term by name.
   *
   * @param query The query string.
   * @param matchType How to match the query.
   * @returns Promise with the response from the server.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async queryOmimTermsByName(
    query: string,
    matchType: string = 'contains',
    ignoreCase: boolean = true
  ): Promise<HpoOmimsResult> {
    const url = `${this.apiBaseUrl}/api/v1/hpo/omims?name=${query}&match=${matchType}&ignore_case=${ignoreCase}`
    const response = await fetch(url, {
      method: 'GET'
    })

    if (!response.ok) {
      const errorBody = await response.json()
      throw new StatusCodeNotOk(errorBody.msg || response.statusText)
    }

    try {
      const responseJson = await response.json()
      return HpoOmimsResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse OMIM response: ${e}`)
    }
  }

  /**
   * Resolve HPO term by ID.
   *
   * @param id HPO ID to resolve.
   * @returns Promise with the resolution result.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async resolveHpoTermById(id: string): Promise<HpoTermResult> {
    const url = `${this.apiBaseUrl}/api/v1/hpo/terms?term_id=${id}`
    const response = await fetch(url, {
      method: 'GET'
    })

    if (!response.ok) {
      // TODO viguno does not responde with a json body on 500 so we don't have a message.
      // TODO once this is fixed in viguno, we can use the code below.
      // TODO https://github.com/varfish-org/viguno/issues/186
      // const errorBody = await response.json()
      // throw new StatusCodeNotOk(errorBody.msg || response.statusText)
      throw new StatusCodeNotOk('HPO term not found')
    }

    try {
      const responseJson = await response.json()
      return HpoTermResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse HPO response: ${e}`)
    }
  }

  /**
   * Query HPO term by name.
   *
   * @param query The query string.
   * @returns Promise with the response from the server.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async queryHpoTermsByName(query: string): Promise<HpoTermResult> {
    const url = `${this.apiBaseUrl}/api/v1/hpo/terms?name=${query}`
    const response = await fetch(url, {
      method: 'GET'
    })

    if (!response.ok) {
      const errorBody = await response.json()
      throw new StatusCodeNotOk(errorBody.msg || response.statusText)
    }

    try {
      const responseJson = await response.json()
      return HpoTermResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse HPO response: ${e}`)
    }
  }

  /**
   * Retrieves HPO terms associated with a gene.
   *
   * @param hgncId HGNC ID of the gene to query for.
   * @returns Promise with the response from the server.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async fetchHpoTermsForHgncId(hgncId: string): Promise<HpoGenesResult> {
    const url = `${this.apiBaseUrl}/api/v1/hpo/genes?gene_id=${hgncId}&hpo_terms=true`
    const response = await fetch(url, {
      method: 'GET'
    })

    if (!response.ok) {
      const errorBody = await response.json()
      throw new StatusCodeNotOk(errorBody.msg || response.statusText)
    }

    try {
      const responseJson = await response.json()
      return HpoGenesResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse HPO response: ${e}`)
    }
  }
}
