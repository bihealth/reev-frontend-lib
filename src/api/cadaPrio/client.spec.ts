import fs from 'fs'
import path from 'path'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

import { CadaPrioClient } from './client'

/** Fixture with prediction results. */
const cadaPrioPredictResultJson = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, './fixture.predictResponse.json'), 'utf8')
)

/** Initialize mock for `fetch()`. */
const fetchMocker = createFetchMock(vi)

describe.concurrent('CadaPrioClient', () => {
  beforeEach(() => {
    fetchMocker.enableMocks()
    fetchMocker.resetMocks()
  })

  it('returns the correct result', async () => {
    // arrange:
    fetchMocker.mockResponseOnce(JSON.stringify(cadaPrioPredictResultJson))

    // act:
    const client = new CadaPrioClient()
    const result = await client.predictGeneImpact(['HP:0000001'])

    // assert:
    expect(JSON.stringify(result)).toMatchSnapshot()
  })

  it('throws in case of fetching problems', async () => {
    // arrange:
    fetchMocker.mockResponse(() => {
      return Promise.reject(new Error('failed to run cada-prio'))
    })

    // act:
    const client = new CadaPrioClient()
    // (with guard)
    await expect(async () => await client.predictGeneImpact(['HP:0000001'])).rejects.toThrow(
      'failed to run cada-prio'
    )

    // assert:
  })
})
