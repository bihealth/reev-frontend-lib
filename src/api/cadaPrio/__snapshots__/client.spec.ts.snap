// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CadaPrioClient > throws in case of fetching problems 1`] = `"{"entries":[{"hgncId":"HGNC:12786","ncbiGeneId":"7476","geneSymbol":"WNT7A","rank":1,"score":76.84823608398438},{"hgncId":"HGNC:11098","ncbiGeneId":"6595","geneSymbol":"SMARCA2","rank":2,"score":76.46592712402344},{"hgncId":"HGNC:25789","ncbiGeneId":"79631","geneSymbol":"EFL1","rank":3,"score":69.20905303955078},{"hgncId":"HGNC:11301","ncbiGeneId":"6729","geneSymbol":"SRP54","rank":4,"score":65.58743286132812},{"hgncId":"HGNC:5154","ncbiGeneId":"3248","geneSymbol":"HPGD","rank":5,"score":63.857845306396484},{"hgncId":"HGNC:4122","ncbiGeneId":"2588","geneSymbol":"GALNS","rank":6,"score":63.840911865234375},{"hgncId":"HGNC:27030","ncbiGeneId":"134218","geneSymbol":"DNAJC21","rank":7,"score":63.35689926147461},{"hgncId":"HGNC:7866","ncbiGeneId":"9241","geneSymbol":"NOG","rank":8,"score":63.207393646240234},{"hgncId":"HGNC:27230","ncbiGeneId":"157570","geneSymbol":"ESCO2","rank":9,"score":63.09525680541992},{"hgncId":"HGNC:2186","ncbiGeneId":"1301","geneSymbol":"COL11A1","rank":10,"score":61.24506378173828},{"hgncId":"HGNC:9587","ncbiGeneId":"9791","geneSymbol":"PTDSS1","rank":11,"score":60.995365142822266},{"hgncId":"HGNC:25507","ncbiGeneId":"55697","geneSymbol":"VAC14","rank":12,"score":60.21630859375},{"hgncId":"HGNC:11602","ncbiGeneId":"6926","geneSymbol":"TBX3","rank":13,"score":59.911136627197266},{"hgncId":"HGNC:4573","ncbiGeneId":"2892","geneSymbol":"GRIA3","rank":14,"score":59.312957763671875},{"hgncId":"HGNC:34016","ncbiGeneId":"100151683","geneSymbol":"RNU4ATAC","rank":15,"score":59.29935836791992},{"hgncId":"HGNC:4220","ncbiGeneId":"8200","geneSymbol":"GDF5","rank":16,"score":58.710994720458984},{"hgncId":"HGNC:23068","ncbiGeneId":"6399","geneSymbol":"TRAPPC2","rank":17,"score":56.57151412963867},{"hgncId":"HGNC:5956","ncbiGeneId":"3549","geneSymbol":"IHH","rank":18,"score":55.67450714111328},{"hgncId":"HGNC:16068","ncbiGeneId":"5116","geneSymbol":"PCNT","rank":19,"score":55.3505859375}]}"`;
